plugins/database/redshift/redshift.go

```diff
@@ -0,0 +1,519 @@
+package redshift
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/database/helper/connutil"
+	"github.com/hashicorp/vault/sdk/database/helper/credsutil"
+	"github.com/hashicorp/vault/sdk/database/helper/dbutil"
+	"github.com/hashicorp/vault/sdk/helper/dbtxn"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	"github.com/lib/pq"
+)
+
+const (
+	sqlTypeName     = "postgres"
+	defaultRenewSQL = `
+ALTER USER "{{name}}" VALID UNTIL '{{expiration}}';
+`
+	defaultRotateRootCredentialsSQL = `
+ALTER USER "{{username}}" WITH PASSWORD '{{password}}';
+`
+
+	defaultRotateCredentialsSQL = `
+ALTER USER "{{name}}" WITH PASSWORD '{{password}}';
+`
+)
+
+var _ dbplugin.Database = &RedShift{}
+
+// New implements builtinplugins.BuiltinFactory
+func New(lowercaseUsername bool) func() (interface{}, error) {
```

Comment:
Great question! `lowercaseUsername` is actually the underlying cause of writing this entire plugin. Redshift implements (mostly) a postgres 8 interface (which is quite old), and part of that is under the hood, it's lowercasing the usernames (the exact opposite of the behavior we saw in Oracle, interestingly). 

Initially, @briankassouf thought we could get away with adding this to our PG client code, shared between many of our secrets engines. In the end, it required us to do that, plus build a specific codebase capable of handling the PG 8 requirements (table names and some SQL verbs varied slightly) of Redshift.

Reply:
Sounds good. It might be useful to add a comment here to that effect so people understand the history and reasoning. I do like the explicitness of keeping the variable, as well.